### PR TITLE
constants in theano rnn; mask in time distributed

### PR DESCRIPTION
**Pull request by braingineer**
In reference to #2376 , this adds the constants to the unmasked version of unrolled theano rnns.  Also, because TimeDistributed is calling the layer directly, it wasn't passing in a mask.  added this as well.

### CODE REVIEW

File: keras/layers/wrappers.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def step(x, states):
             # to process batches of any size
             # we can go with reshape-based implementation for performance
             X = K.reshape(X, (-1, ) + input_shape[2:])  # (nb_samples * timesteps, ...)
-            y = self.layer.call(X)  # (nb_samples * timesteps, ...)
+            y = self.layer.call(X, mask)  # (nb_samples * timesteps, ...)
```
Comment (fchollet) [inline keras/layers/wrappers.py]: This won't work, since the shape of the mask won't match the shape of the input.

**Comment by braingineer** [comment]
actually, now that I think about it.. the mask in TimeDistributed might need to be handled differently than just putting it in there..

File: keras/layers/wrappers.py (hunk 1 above)
Comment (fchollet) [inline keras/layers/wrappers.py]: The proper solution in this case is to add a test `if mask is None` and go with the `rnn`-based solution above in that case.

Comment (braingineer) [inline keras/layers/wrappers.py]: do you mean something like:

```python
        if input_shape[0] or mask is not None:
            # batch size matters, use rnn-based implementation
            def step(x, states):
                output = self.layer.call(x)
                return output, []

            last_output, outputs, states = K.rnn(step, X, mask=mask,
                                                 initial_states=[])
            y = outputs
```

Comment (fchollet) [inline keras/layers/wrappers.py]: Almost, but not quite. There is one more thing to consider, which is that `mask` here is for masking the timesteps of the current layer, not that of the child layer. Fundamentally that's the reason why `TimeDistributed` isn't using masking in its current implementation.

Comment (braingineer) [inline keras/layers/wrappers.py]: Isn't that already handled by the rnn? 

Example case:
```python
xin = Input(batch_shape=(batch, time), dtype='int32')
xemb = Embedding(feat, embed)(xin)
xrnn = LSTM(rnn_size, return_sequences=True)(x_emb)
xout = TimeDistributed(Dense(feat, activation='softmax'))(xrnn)
```
- input shape: `(batch, time)`
- post embedded shape: `(batch, time, embed)`
- post lstm shape (with return sequences):` (batch, time, rnn_size)`
- then, one would want to apply a `TimeDistributed(Dense(...))` over this tensor. 

then

- The mask that came out of the embedding is 2dim (it's the same size as the input)
- The mask gets pushed through `xrnn` as is
- Then, the mask hits the `TimeDistributed`, of shape` (batch, time)`.  

next,

- the mask gets passed to `TimeDistributed`'s rnn, and reshaped to `(time, batch)`.  
- in the `theano_backend. rnn`, it does `output = T.switch(mask, output, output_tm1)`, 
- each step through the rnn is size `(batch,)`.  is this right? 
- I am assuming that this means the batch examples which were 0'ed out and the time steps which shouldn't be calculated are properly handled.  Is this right?

If I'm wrong on any part of that, please let me know.

**Comment by fchollet** [comment]
Please remove the part related to mask in TimeDistributed. I'll merge the rnn fix.

**Comment by braingineer** [comment]
somehow something is weird. I'm just going to reset the branch and do another merge request.
